keps/sig-node/4580-deprecate-kubelet-runonce/README.md

```diff
@@ -0,0 +1,355 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Deprecated and removed kubelet support for RunOnce mode.
+
+## Motivation
+
+* Runonce does not work in systemd mode.
+* Runonce mode doesn't support many newer pod features (init containers).
+* The pod lifecycle for runonce mode is even less well defined than normal.
+* The mode appears to be entirely untested in CI, except for a few unit tests.
+
+### Goals
+
+* Mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated,  and finally remove the runonce flag.
+* Remove kubelet support for RunOnce mode.
+
+### Non-Goals
+
+Immediate removal: the deprecation and removal process will be gradual and feature gate to increase awareness among potential users.
+
+## Proposal
+
+The runonce mode of kubelet will exit the kubelet process after spawning pods from the local manifests or remote URL. It is suitable for scenarios where one-time tasks need to be run on the node, this proposal outlines plans to deprecate and remove runonce mode in kubelet.
+
+### Risks and Mitigations
+
+Some people may still rely on this feature, but podman addresses the same use case with more well-supported way, ref: https://docs.podman.io/en/latest/markdown/podman-kube.1.html. Affected users can migrate to *podman kube subcommand* on demand.
+
+## Design Details
+
+### KubeletConfiguration Change: KubeletConfiguration
+
+Mark the `RunOnce` field as deprecated.
+
+### kubelet flag Change
+
+make the `runonce` flag as deprecated, and remove it in GA version.
+
+### Implement warning logging for runonce mode usage
+
+Starting in 1.31, during kubelet startup, if running in runonce mode, the kubelet will log a warning message, for example:
+
+```
+klog.Warning("RunOnce mode has been deprecated, and will be removed in a future release")
+```
+
+### Introduction LegacyNodeRunOnceMode feature gate
+
+With the introduction of the `LegacyNodeRunOnceMode` feature gate, Kubernetes aims to guide users through the deprecated runonce mode. When this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.
```

Comment:
`LegacyNodeRunOnceMode` suggests the opposite of what you're proposing here. Iow. setting `LegacyNodeRunOnceMode=true` it will not allow using the deprecated RunOnce mode. Maybe something like `RemovedNodeRunOnceMode` or something that explicitly says it's not there will be a better feature gate name?

Reply:
The reason for using Legacy* is that we have used it in some areas...
ref: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates-removed/

Reply:
We can invert the sense of the feature gate: ie, you disable it to test running without the deprecated mode, and the feature gate turns false-by-default before we eventually remove it.

Reply:
Try:
```diff
-Once this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.
+Unless this feature gate is enabled, kubelet will refuse to start when the `--runonce` command line flag is set.
```

Reply:
I reversed it wherever it was mentioned (hopefully this won't get any worse).